### PR TITLE
feat(routing): add support for multiple path segment fields

### LINKED ISSUE #423: Add support for filters in route templates
This is a request I've heard from the community, and that was previously mentioned in one of my comments on #161.  It would be nice to if we could support this within the path and for query string params (assuming we implement the latter). We can take inspiration from [bottle](http://bottlepy.org/docs/dev/tutorial.html#dynamic-routes). Default filters might include:
- [x] `int` (already implemented)
- [x] `float` deferred to https://github.com/falconry/falcon/issues/2022
- [ ] `path` prototype PR: https://github.com/falconry/falcon/pull/1945
- [x] `re` deferred to https://github.com/falconry/falcon/issues/857
- [ ] `in` will be deferred to another issue if needed

~~Also, we need to consider how this might be applies to query string parameters as well (which would require defining such parameters in the URI template)~~ :arrow_left: tracked separately in https://github.com/falconry/falcon/issues/838

### LINKED ISSUE #648: Routing all paths to one handler function
Is it possible to route all paths that start with something like “/api” to same handler function?

Such as:

```
/api/foo
/api/bar
/api/foo/bar
/api/bar/baz/this/that 
```

All should be handled with one function and I should be able to get the full path after /api.

This feature is very handy and I used it often at Node.js Express framework. Now I am looking for ways to accomplish the same thing with Falcon.

More info can be found here; It defines the feature as "white-listed “global” functionality."
http://expressjs.com/api.html#app.all

### PR DESCRIPTION
**Pull request by CaselIT**
# Summary of Changes

This adds support for a converter to be declared as consuming all the remaining path.

The current implementation does not allow nesting these converters and is quite conservative (I think), but I believe we can improve on this later if we need to.

# Related Issues

Closes #423
Fixes #648
Relates to #1895

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [x] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [x] Added **tests** for changed code.
- [x] Prefixed code comments with GitHub nick and an appropriate prefix.
- [x] Coding style is consistent with the rest of the framework.
- [x] Updated **documentation** for changed code.
    - [ ] Added docstrings for any new classes, functions, or modules.
    - [ ] Updated docstrings for any modifications to existing code.
    - [ ] Updated both WSGI and ASGI docs (where applicable).
    - [ ] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [ ] Updated all relevant supporting documentation files under `docs/`.
    - [ ] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [ ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [ ] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/actual-freaking-docs/index.html) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

### CODE REVIEW

File: tests/asgi/test_ws.py
Comment (CaselIT) [inline tests/asgi/test_ws.py:345]: this fixes the warning 
`The explicit passing of coroutine objects to asyncio.wait() is deprecated since Python 3.8, and scheduled for removal in Python 3.11.`

Comment (CaselIT) [reply tests/asgi/test_ws.py:345]: this hangs in py36 -.-'

Comment (CaselIT) [reply tests/asgi/test_ws.py]: fixed now

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:34]: this should probably renamed in `consume_multiple_segments` or similar so that it keeps making sense if in the future we enable support for `prefix/{rest:path}/suffix`

**Comment by CaselIT** [comment]
After https://github.com/falconry/falcon/pull/1990 is merged we should update the docs there to mention that for simple cases `:path` may be used

**Comment by CaselIT** [comment]
I'm not really sold on the idea, since it's another thing that should be documented, explained etc.

I think your example would work better with `/route/{itemid}/repository/{url:relativepath}`. `path` is not special, meaning that the way it indicates that it can captures more than one segment is available to other converters

**Comment by vytas7** [comment]
Agreed that my idea ain't very pretty either.
But functionality wise, it depends on the way you view it. Say that you have a useful converter that _might_ accept path, but in another case you would want to restrict it to a single field, or just use it in another field than the very last one (suffix). I guess you could subclass it to override that flag, and then register the new type as well, but it won't be particularly pretty to document either.

**Comment by CaselIT** [comment]
Since the interface of the converters changes, I think it makes sense to have two implementation in that case. (path like converters get passed a list of path segments, not a single segment)

**Comment by vytas7** [comment]
Another issue in that useful to have in mind when discussing the big picture: https://github.com/falconry/falcon/issues/1567 (which would actually favour your current design, but maybe we could devise a more generic way to provide hints?)

**Comment by CaselIT** [comment]
From a quick look I don't think that the current implementation is incompatible with that issue

**Comment by vytas7** [comment]
I know that although Falcon has moved away from [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570), the said paper has all kinds of funky expansions, so maybe something familiar could be adapted from [Section 3.2.6](https://datatracker.ietf.org/doc/html/rfc6570#section-3.2.6), like `{list*}` :arrow_right: `/red/green/blue`.

Or maybe we could have a discussion [meeting?] about this on chat at some point? I don't have any elegant suggestions, but maybe someone else does.

**Comment by CaselIT** [comment]
That may be an option, but we right having two ways of doing something, but having both the converters and this special syntax.
Also what would happen if both a converter and `*` are used together? Like `{list*:int}`

**Comment by vytas7** [comment]
Dunno, I was just thinking what the options were before we move forward with this. But we should move forward in one or another way 🙂

**Comment by vytas7** [comment]
It would be also good to reason about this change in the perspective of adding `re` converter (https://github.com/falconry/falcon/issues/857). Requiring a `CONSUME_PATH` attribute on the converter class would force a decision to make `re` of one or another type, however, I can imagine both could be useful :thinking:

File: falcon/routing/compiled.py
Comment (kgriffs) [inline falcon/routing/compiled.py:276]: Why do we `nodes.append(new_node)` in both branches of the `if...else` as opposed to doing it once outside the statement?

File: falcon/routing/compiled.py
Comment (kgriffs) [inline falcon/routing/compiled.py:553]: nice sanity check

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:158]: `matches the all the remaining url path` -->  `matches the remainder of the URL path`

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:159]: We should also note what the reader should expect when the path converter comes at the end of a path that is already consumed. For example, given:

Template: `'/foo/bar/{foo:path}'`
Path Requested: `'/foo/bar'`

The path will not match. Also describe what happens when there's a trailing forward-slash in the requested path?

`'/foo/bar/'`

File: tests/test_default_router.py
Comment (kgriffs) [inline tests/test_default_router.py:781]: We should also test that the following request paths result in 404s when `'/{baz:path}'` is *not* added as a route:

`'/foo'`
`'/foo/'`
`'/foo/bar'`
`'/foo/bar/'`

**Review by kgriffs — CHANGES_REQUESTED** [review]

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py:276]: I guess the idea was to add it only once the checks are passed since raising an error after may lead to an inconsistent router state?

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py:240]: The idea is that this limitation is temporary, removing the limitation once that's done. Also in other places the router already raises a ValueError

**Comment by CaselIT** [comment]
thanks for the review @kgriffs 

sorry for the delay in getting to this

**Comment by CaselIT** [comment]
Btw I've taken another look at supporting additional segments after a "consume multiple segments" converter it's not trivial.
Also there is a decision to take here: should a "consume multiple segments" converter be allowed to match 0 segments?

Assuming the reply is yes, the finder would need to be something like the following
```py
from dataclasses import dataclass, field
from itertools import count
from falcon.routing import CompiledRouter
from falcon.routing.converters import PathConverter

seq = count(1)

@dataclass
class Tmp:
    id: int = field(default_factory=seq.__next__)

    def on_get(self, req, resp, **kw):
        resp.media = {'id': self.id, 'kw': kw}

r = CompiledRouter()

def add(url, res):
    print(res, url.replace('{x}', '{x:path}'))
    r.add_route(url, res)

add('/foo/{x}', Tmp())
add('/foo/{x}/foo', Tmp())
add('/foo/{x}/{baz:int}', Tmp())
add('/foo/{x}/foo/bar', Tmp())
add('/foo/{x}/{baz:int}/baz', Tmp())

r._compile()
# print(r.finder_src)

def find(path, return_values, patterns, converters, params):
    converters = [PathConverter(), *converters]
    rv = {r.resource.id: r for r in return_values}
    path_len = len(path)
    if path_len > 0:
        if path[0] == 'foo':
            if path_len > 1:
                remaining_path = path[1:]
                remaining_path_len = len(remaining_path)
                if remaining_path_len > 1:
                    if remaining_path[-1] == 'bar':
                        if remaining_path[-2] == 'foo':
                            fragment = remaining_path[:-2]
                            field_value_1 = converters[0].convert(fragment)
                            if field_value_1 is not None:
                                params['x'] = field_value_1
                                return rv[4]
                    if remaining_path[-1] == 'baz':
                        fragment = remaining_path[-2]
                        field_value_2 = converters[1].convert(fragment)
                        if field_value_2 is not None:
                            fragment = remaining_path[:-2]
                            field_value_1 = converters[0].convert(fragment)
                            if field_value_1 is not None:
                                params['baz'] = field_value_2
                                params['x'] = field_value_1
                                return rv[5]
                if remaining_path_len > 0:
                    if remaining_path[-1] == 'foo':
                        fragment = remaining_path[:-1]
                        field_value_1 = converters[0].convert(fragment)
                        if field_value_1 is not None:
                            params['x'] = field_value_1
                            return rv[2]
                    fragment = remaining_path[-1]
                    field_value_2 = converters[1].convert(fragment)
                    if field_value_2 is not None:
                        fragment = remaining_path[:-1]
                        field_value_1 = converters[0].convert(fragment)
                        if field_value_1 is not None:
                            params['baz'] = field_value_2
                            params['x'] = field_value_1
                            return rv[3]
                field_value_1 = converters[0].convert(remaining_path)
                if field_value_1 is not None:
                    params['x'] = field_value_1
                    return rv[1]
                return None
            return None
        return None
    return None

r._find = find

def go(url):
    res = r.find(url)
    if res is None:
        print(url, res)
    resource, _, param, _ = res
    print(resource, param, url)

print('-' * 20)
go('/foo/foo')
go('/foo/bar')
go('/foo/bar/123')
go('/foo/bar/123/x')
go('/foo/bar/123/x/foo')
go('/foo/foo/bar/123/x/foo')
go('/foo/bar/123/x/foo/bar')
go('/foo/123/baz')
go('/foo/bar/123/baz')
go('/foo/bar/123/x/foo/bar/123/baz')
```

running the above yields
```
Tmp(id=1) /foo/{x:path}
Tmp(id=2) /foo/{x:path}/foo
Tmp(id=3) /foo/{x:path}/{baz:int}
Tmp(id=4) /foo/{x:path}/foo/bar
Tmp(id=5) /foo/{x:path}/{baz:int}/baz
--------------------
Tmp(id=2) {'x': ''} /foo/foo
Tmp(id=1) {'x': 'bar'} /foo/bar
Tmp(id=3) {'baz': 123, 'x': 'bar'} /foo/bar/123
Tmp(id=1) {'x': 'bar/123/x'} /foo/bar/123/x
Tmp(id=2) {'x': 'bar/123/x'} /foo/bar/123/x/foo
Tmp(id=2) {'x': 'foo/bar/123/x'} /foo/foo/bar/123/x/foo
Tmp(id=4) {'x': 'bar/123/x'} /foo/bar/123/x/foo/bar
Tmp(id=5) {'baz': 123, 'x': ''} /foo/123/baz
Tmp(id=5) {'baz': 123, 'x': 'bar'} /foo/bar/123/baz
Tmp(id=5) {'baz': 123, 'x': 'bar/123/x/foo/bar'} /foo/bar/123/x/foo/bar/123/baz
```

In the above example `/foo/{x:path}/foo` matches `/foo/foo` with `x=''`

I'll copy the above in a new issue

**Comment by kgriffs** [comment]
> There is a decision to take here: should a "consume multiple segments" converter be allowed to match 0 segments? the current implementation allows it.

Personally I would expect it to never match if there is no path to consume, because otherwise what's the point of using the `path`	converter? Or at least, if there is another route registered for the exact path prefix, that one would take precedence over the route containing the `path` converter.

One could argue it my be useful to be able to handle `/foo/bar/bang` and /foo/bang` within the same resource controller, but that doesn't feel like a good API design to me (though it may be useful for re-implementing a legacy API?). So perhaps we could introduce an option to allow 0 segments but it is disabled by default? Or we introduce that along with #2061 ?

**Comment by kgriffs** [comment]
Code looks good, we just need a consensus on what to do about allowing 0-length path matches.

**Comment by vytas7** [comment]
If I understand the problem right, IMHO, for the sake of consistency, the new converter should behave like fields behave now.
They do match an empty segment after a trailing slash, and they do not otherwise.

Given the following test app,
```python
import falcon

class Resource:
    def on_get(self, req, resp, resourceid):
        resp.media = {'path': req.path, 'resourceid': resourceid}

(app := falcon.App()).add_route('/api/{resourceid}', Resource())
```

Cf
```
$ xh http://localhost:8000/api/
HTTP/1.1 200 OK
Content-Length: 35
Content-Type: application/json

{
    "path": "/api/",
    "resourceid": ""
}
```
vs
```
$ xh http://localhost:8000/api
HTTP/1.1 404 Not Found
Content-Length: 26
Content-Type: application/json
Vary: Accept

{
    "title": "404 Not Found"
}
```

**Comment by CaselIT** [comment]
That's the current behavior, unless `strip_url_path_trailing_slash=True`, in which case both yield 404

Also note that my question was only once we add support for additional parts after the path. Basically the question what if
`/foo/{bar:path/baz` should match `/foo/baz` or not.

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py:271]: good idea, will add a comment

**Comment by vytas7** [comment]
Aha, right, I didn't realize question was only about when we lift the requirement for a `path` segment to be last.
Then maybe we don't need to make a decision _right now_?

In any case, I would perceive routing as strange and confusing if `/foo/{bar:path}/baz` matched `/foo/baz`, it wouldn't be consistent with the current behaviour where a field essentially needs a preceding `/` to match. So my vote is it shouldn't match.

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py]: I've created a subclass and use it for each error in add_route

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py:271]: In the end it's probably better to just add the node and remove if before raising an error if needed

**Comment by CaselIT** [comment]
> Aha, right, I didn't realize the question was only about when we lift the requirement for a `path` segment to be last. Then maybe we don't need to make a decision _right now_?
> 
> In any case, I would perceive routing as strange and confusing if `/foo/{bar:path}/baz` matched `/foo/baz`, it wouldn't be consistent with the current behaviour where a field essentially needs a preceding `/` to match, and in this case the `path` field would rob `/baz` of that slash. So my vote is on "it shouldn't match".

We really ought to move the conversation here https://github.com/falconry/falcon/issues/2061

**Comment by vytas7** [comment]
Sure -- I was just pinged here on this issue :slightly_smiling_face:

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:167]: Given this is documented for the developer, I'm OK with leaving the behavior as-is.

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py]: I've updated, let me know what you think

**Comment by kgriffs** [comment]
> If I understand the problem right, IMHO, for the sake of consistency, the new converter should behave like fields behave now. They do match an empty segment after a trailing slash, and they do not otherwise.

Good point, I'm OK with moving forward w/ this as long as the behavior is explicitly documented for the developer.

> the question was only about when we lift the requirement for a path segment to be last.

Actually, I wanted your input on both questions. 😄 Second one is not urgent. Thanks!

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:35]: `converted` --> `converter`

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:38]: `all the remaining url path` --> `all of the remaining URL path segments`

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:36]: capitalize `URL` here and in the next line since it is an acronym.

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:165]: `url` --> `URL`

File: falcon/routing/converters.py
Comment (kgriffs) [inline falcon/routing/converters.py:171]: Perhaps also mention that `False` is the default for this option.

**Review by kgriffs — APPROVED** [review]

**Comment by vytas7** [comment]
Yes so, I think no converter should match 0 fields, because that would be a substantial change in comparison to how the framework behaves now. Matching an empty string after a trailing slash is a bit special edge case (or oversight) which already exists now, so path converter can also do the same, but `/foo/{bar}/{baz:path}` ought not to match `/foo/bar`.

File: docs/api/routing.rst
Comment (vytas7) [inline docs/api/routing.rst:298]: It seems that ``prefix/{other:path}`` is misaligned

File: docs/_newsfragments/648.rst
Comment (vytas7) [inline docs/_newsfragments/648.rst:1]: In newsfragments, IMHO we should tell a story, not phrase in the imperative as docstrings.
Also, _converted_ (sic) :arrow_right: _converter_.

File: falcon/routing/converters.py
Comment (vytas7) [inline falcon/routing/converters.py:61]: Cosmetic: `BaseConverter` would look nicer as a class reference.

File: falcon/routing/compiled.py
Comment (vytas7) [inline falcon/routing/compiled.py:214]: Do we really need to do that patching? We could just do a `getattr` here instead.

File: falcon/routing/converters.py
Comment (vytas7) [inline falcon/routing/converters.py:173]: It might be useful to link to ["How does Falcon handle a trailing slash in the request path?"](https://falcon.readthedocs.io/en/stable/user/faq.html#how-does-falcon-handle-a-trailing-slash-in-the-request-path) from here.

**Review by vytas7 — CHANGES_REQUESTED** [review]
This is shaping up, well done :+1:

A couple of inline nitpicks aside, we also ought to revise this FAQ item: [How can I handle forward slashes within a route template field?](https://falcon.readthedocs.io/en/stable/user/faq.html#how-can-i-handle-forward-slashes-within-a-route-template-field)

**Comment by CaselIT** [comment]
thanks for the review

> but `/foo/{bar}/{baz:path}` ought not to match `/foo/bar`.

it does not at the moment.
without `/` in the last segment the behavior will be the same as if using  `/foo/{bar}/{baz}` as template

**Comment by kgriffs** [comment]
> [vytas7](https://github.com/vytas7) requested changes [20 hours ago](https://github.com/falconry/falcon/pull/1945#pullrequestreview-971407378)
lol, this is why it is good to require >= 2 approvals. You guys inevitably catch things I miss.

File: falcon/routing/compiled.py
Comment (CaselIT) [reply falcon/routing/compiled.py:214]: no clue why I imokemented that way, but it seems we could avoid that

I think the idea was to have a place that could be used to normalize the api of the converters, in case in the future we wanted to do some other vhange to them.

but it does indeed seem overkill for this case

File: docs/_newsfragments/648.rst
Comment (CaselIT) [reply docs/_newsfragments/648.rst]: tried rewording. I'm afraid I'm not a story teller, so feel free to suggest an alternative

File: docs/_newsfragments/648.newandimproved.rst
Comment (vytas7) [inline docs/_newsfragments/648.newandimproved.rst:1]: Would be nice to link to the new converter class from here, although not required to pass review.
For example, the RST code could read:
```rst
A new ``path`` :class:`converter <~falcon.routing.PathConverter>` capable of matching segments that include ``/``...
```

**Review by vytas7 — CHANGES_REQUESTED** [review]
LGTM sans the still pending FAQ update: [How can I handle forward slashes within a route template field?](https://falcon.readthedocs.io/en/stable/user/faq.html#how-can-i-handle-forward-slashes-within-a-route-template-field)

File: docs/_newsfragments/648.newandimproved.rst
Comment (CaselIT) [reply docs/_newsfragments/648.newandimproved.rst:1]: will use that

**Comment by CaselIT** [comment]
> LGTM sans the still pending FAQ update: [How can I handle forward slashes within a route template field?](https://falcon.readthedocs.io/en/stable/user/faq.html#how-can-i-handle-forward-slashes-within-a-route-template-field)

sorry I missed the previous message. Will do that

**Review by vytas7 — APPROVED** [review]
LGTM :+1:

**Pull request merged**
